src/org/opendatakit/validate/FormValidator.java

```diff
@@ -324,28 +311,81 @@ boolean stepThroughEntireForm(FormEntryModel model) throws InvalidReferenceExcep
         return outcome;
     }
 
-    void validate(String path) {
+    public void validateAndExitWithErrorCode(String path) {
+        try {
+            validate(path);
+        } catch (Exception e) {
+            errors.error("\nException: ", e);
+            setError(true);
+        }
 
-    	File src = new File(path);
-    	if ( !src.exists() ) {
-    		setError(true);
-    		System.err.println("File: " + src.getAbsolutePath() + " does not exist.");
-    		return;
-    	}
+        if (inError) {
+            errors.error("\nResult: Invalid");
+            System.exit(1);
+        } else {
+            System.exit(0);
+        }
+    }
+
+  public void validate(String path) {
+        File src = new File(path);
+        if (!src.exists()) {
+            setError(true);
+            errors.error("File: " + src.getAbsolutePath() + " does not exist.");
+            return;
+        }
 
         FileInputStream fis = null;
         try {
             fis = new FileInputStream(src);
+            validate(fis);
+        } catch (FileNotFoundException e) {
+            setError(true);
+            errors.error("Please choose a file before attempting to validate.");
+            return;
+        } finally {
+            if (fis != null) {
+                try {
+                    fis.close();
+                } catch (IOException e) {
+                    // ignore
+                    e.printStackTrace();
+                }
+            }
+        }
 
-	        // validate well formed xml
-	        // System.out.println("Checking form...");
+    }
+
+
+
+    public void validateText(String xml) {
+        validate(new ByteArrayInputStream(xml.getBytes()));
+    }
+
+
+    public void validate(InputStream xmlSource) {
+
+            byte[] xformData;
+            try {
+                // first read the whole form into memory since this stream is going to be read twice
+                // first for xml validation
+                // second for xform validation
+                ByteArrayOutputStream baos = copyToByteArray(xmlSource);
```

Comment:
If I understand correctly, this is an attempt at improving performance by only reading the file into memory once. Does it actually improve performance? I believe the overall memory footprint is bigger and the code is less readable so I'd like to know how you've quantified that it's an improvement.

Also, since I don't think this is related to the `ErrorListener` interface, it would have been better as a separate PR. For next time. 😊